src/components/AutoComplete.tsx

```diff
@@ -0,0 +1,288 @@
+import React, { createContext, useContext, useCallback, useMemo, useEffect, useState, useRef } from 'react';
+import styled from 'styled-components';
+import { gapS, gapXs, gray4, gray7 } from '@taskany/colors';
+
+import { nullable } from '../utils';
+
+import { Text } from './Text';
+import { Input } from './Input';
+
+type InputProps = React.ComponentProps<typeof Input>;
+type AutoCompleteMode = 'single' | 'multiple';
+type AutoCompleteSelectedMap<T> = Set<T>;
+
+interface AutoCompleteRenderItemProps<T> {
+    item: T;
+    index: number;
+    onItemClick: () => void;
+    checked: boolean;
+}
+
+interface AutoCompleteRenderItem<T> {
+    (props: AutoCompleteRenderItemProps<T>): React.ReactNode;
+}
+
+interface AutoCompleteContext<T> {
+    items: T[];
+    value: T[];
+    renderItem: AutoCompleteRenderItem<T>;
+    renderItems: (props: React.PropsWithChildren) => React.ReactNode;
+    renderState: 'combine' | 'split';
+    switchType: () => void;
+    popItem: (item: T) => void;
+    pushItem: (item: T) => void;
+    map: React.MutableRefObject<AutoCompleteSelectedMap<T>>;
+}
+
+interface AutoCompleteProps<T> {
+    mode: AutoCompleteMode;
+    items: T[];
+    renderItem: AutoCompleteRenderItem<T>;
+    renderItems?: (props: React.PropsWithChildren) => React.ReactNode;
+    value?: T[];
+    onChange: (items: T[]) => void;
+}
+
+interface AutoCompleteListProps {
+    title?: string;
+    selected?: boolean;
+}
+
+interface AutoCompleteInputProps extends Omit<InputProps, 'onChange'> {
+    onChange: (val: string) => void;
+}
+
+interface AutoCompleteRadioGroupProps<T extends { title: string; value: string }> {
+    name: string;
+    title: string;
+    items: T[];
+    value?: T['value'];
+    onChange: (value: T) => void;
+    className?: string;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const AutoCompleteContextProvider = createContext<AutoCompleteContext<any> | null>(null);
+
+export function useAutoCompleteContext<T extends { id: string }>(): AutoCompleteContext<T> {
```

Comment:
Any cases to use it outside?

Reply:
dunno, this check able exclude nullish value for hook's result